In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.grid_search import GridSearchCV
from sklearn.linear_model import LogisticRegression

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


## Reading Avg w2v vectors(200-D)

In [2]:
avg_w2v =pd.read_pickle("\\Users\\AKSHAY\\Desktop\\ML python course\\Amazon Project\\Logistic Regression\\W2V Vectors\\Avg W2V Vectors\\avg_w2v_vec_200")
print(avg_w2v.shape)
avg_w2v.head()

(122109, 201)


0         1         2         3         4         5         6    \
0  0.000389 -0.000173  0.000436 -0.000163  0.000199  0.000180  0.000041   
1  0.000284  0.000089  0.000249  0.000084 -0.000130  0.000052 -0.000071   
2 -0.000061 -0.000244  0.000047 -0.000635 -0.000620  0.000108  0.000711   
3  0.000390 -0.000164  0.000224 -0.000373 -0.000065 -0.000066  0.000333   
4  0.000057  0.000510 -0.000419  0.000030 -0.000151 -0.000305 -0.000428   

        7         8         9   ...        191       192       193       194  \
0  0.000549  0.000178  0.000004 ...  -0.000026 -0.000466 -0.000134  0.000429   
1  0.000123 -0.000138 -0.000318 ...  -0.000117 -0.000317 -0.000018 -0.000481   
2 -0.000231 -0.000256  0.000566 ...   0.000208 -0.000340 -0.000010  0.000276   
3 -0.000012 -0.000021  0.000121 ...   0.000292 -0.000301 -0.000081 -0.000084   
4  0.000091  0.000022  0.000173 ...  -0.000019  0.000849 -0.000127  0.000260   

        195       196       197       198       199  200  
0 -0.000480  0.000480 -0.000280 -0.000104  0.000497    1  
1  0.000017 -0.000097  0.000324 -0.000045 -0.000154    1  
2  0.000060  0.000266  0.000313 -0.000156  0.000381    0  
3 -0.000054 -0.000090 -0.000026  0.000211 -0.000205    1  
4 -0.000177  0.000521  0.000027 -0.000096  0.000205    1  

[5 rows x 201 columns]

In [3]:
x =avg_w2v.iloc[:,:200]
y =avg_w2v.iloc[:,200]

In [4]:
print(x.shape)
y.shape

(122109, 200)


(122109,)

## Column Standardization

In [5]:
from sklearn.preprocessing import StandardScaler

In [6]:
a =StandardScaler()
x =a.fit_transform(x)

In [7]:
x_train,x_test,y_train,y_test =train_test_split(x,y,test_size =0.3)

In [8]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
y_test.shape

(85476, 200)
(36633, 200)
(85476,)


(36633,)

In [9]:
tuned_params = [{'C': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100, 1000, 10000]}]

In [10]:
model = GridSearchCV(LogisticRegression(), tuned_params, scoring = 'accuracy')
model.fit(x_train, y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid=[{'C': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100, 1000, 10000]}],
       pre_dispatch='2*n_jobs', refit=True, scoring='accuracy', verbose=0)

In [ ]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

In [12]:
print(model.best_estimator_)
print(model.score(x_test, y_test))

LogisticRegression(C=0.01, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
0.756612890017


In [13]:
y_pred = model.best_estimator_.predict(x_test)

In [14]:
acc = accuracy_score(y_test, y_pred, normalize=True) * float(100)
acc

75.661289001719751

We can clearly see increasing dimenions the model performs better than 100-D as the chances of data becomes linearly separable in higher dimensions increases.So we get the best hyperplane as dimensions increases due to which model behaves better.

## Confusion Matrix

In [16]:
confusion_matrix(y_test,y_pred).T

array([[12493,  4193],
       [ 4723, 15224]], dtype=int64)

## L1 regularization and sparsity

In [18]:
clf = LogisticRegression(C=1, penalty = 'l1')
clf.fit(x_train, y_train)
w = clf.coef_    
print(np.count_nonzero(w))

200


This means 200 out of 200 are non zero.That means for regularization takes place.

In [20]:
clf = LogisticRegression(C=0.1, penalty = 'l1')
clf.fit(x_train, y_train)
w = clf.coef_
print(np.count_nonzero(w))

198


As we can see clearly as c decreases sparsity increases.This time 198 features out of 200  are non zero,means 2 features are removed.

In [22]:
clf = LogisticRegression(C=0.001, penalty = 'l1')
clf.fit(x_train, y_train)
w = clf.coef_
print(np.count_nonzero(w))

94


Now the 94 out of 200 are non zero.That means 106 are zero which are removed.

In [40]:
clf = LogisticRegression(C=0.00025, penalty = 'l1')
clf.fit(x_train, y_train)
w = clf.coef_
print(np.count_nonzero(w))

15


At c=0.00025 we are getting top 15 features and rest becomes zero.